File: backend/apps/company/serializers/company_settings.py
"""
CompanySettings serializers — read and partial update.

CompanySettingsSerializer: full representation for GET responses.
CompanySettingsUpdateSerializer: partial update with field-level validation.
"""

from __future__ import annotations

from rest_framework import serializers

from apps.company.constants import (
    PRECISION_MAX,
    PRECISION_MIN,
)
from apps.company.models import CompanySettings


class CompanySettingsSerializer(serializers.ModelSerializer):
    """Read-only representation of all company settings."""

    class Meta:
        model = CompanySettings
        exclude = ("id",)
        read_only_fields = [field.name for field in CompanySettings._meta.get_fields() if hasattr(field, "column")]


class CompanySettingsUpdateSerializer(serializers.Serializer):
    """Validates partial updates to CompanySettings.

    All fields are optional — only provided fields are updated.
    Model-level cross-field validation (separator conflicts, fiscal day,
    etc.) runs in CompanySettings.clean() at save time.
    """

    # Number formatting
    decimal_separator = serializers.ChoiceField(
        choices=CompanySettings.decimal_separator.field.choices,
        required=False,
    )
    thousands_separator = serializers.ChoiceField(
        choices=CompanySettings.thousands_separator.field.choices,
        required=False,
    )
    digit_grouping = serializers.ChoiceField(
        choices=CompanySettings.digit_grouping.field.choices,
        required=False,
    )

    # Precision
    quantity_precision = serializers.IntegerField(
        min_value=PRECISION_MIN,
        max_value=PRECISION_MAX,
        required=False,
    )
    price_precision = serializers.IntegerField(
        min_value=PRECISION_MIN,
        max_value=PRECISION_MAX,
        required=False,
    )
    currency_precision = serializers.IntegerField(
        min_value=PRECISION_MIN,
        max_value=PRECISION_MAX,
        required=False,
    )
    exchange_rate_precision = serializers.IntegerField(
        min_value=PRECISION_MIN,
        max_value=PRECISION_MAX,
        required=False,
    )
    percentage_precision = serializers.IntegerField(
        min_value=PRECISION_MIN,
        max_value=PRECISION_MAX,
        required=False,
    )
    weight_precision = serializers.IntegerField(
        min_value=PRECISION_MIN,
        max_value=PRECISION_MAX,
        required=False,
    )

    # Currency display
    currency_symbol_position = serializers.ChoiceField(
        choices=CompanySettings.currency_symbol_position.field.choices,
        required=False,
    )
    currency_spacing = serializers.BooleanField(required=False)
    negative_format = serializers.ChoiceField(
        choices=CompanySettings.negative_format.field.choices,
        required=False,
    )
    multi_currency_enabled = serializers.BooleanField(required=False)

    # Rounding
    rounding_method = serializers.ChoiceField(
        choices=CompanySettings.rounding_method.field.choices,
        required=False,
    )
    cash_rounding_enabled = serializers.BooleanField(required=False)
    cash_rounding_increment = serializers.DecimalField(
        max_digits=4,
        decimal_places=2,
        required=False,
    )

    # Date & time
    date_format = serializers.ChoiceField(
        choices=CompanySettings.date_format.field.choices,
        required=False,
    )
    time_format = serializers.ChoiceField(
        choices=CompanySettings.time_format.field.choices,
        required=False,
    )
    week_start_day = serializers.ChoiceField(
        choices=CompanySettings.week_start_day.field.choices,
        required=False,
    )
    calendar_system = serializers.ChoiceField(
        choices=CompanySettings.calendar_system.field.choices,
        required=False,
    )

    # Measurement
    measurement_system = serializers.ChoiceField(
        choices=CompanySettings.measurement_system.field.choices,
        required=False,
    )
    default_weight_uom = serializers.ChoiceField(
        choices=CompanySettings.default_weight_uom.field.choices,
        required=False,
    )
    default_length_uom = serializers.ChoiceField(
        choices=CompanySettings.default_length_uom.field.choices,
        required=False,
    )
    default_volume_uom = serializers.ChoiceField(
        choices=CompanySettings.default_volume_uom.field.choices,
        required=False,
    )
    default_temperature_uom = serializers.ChoiceField(
        choices=CompanySettings.default_temperature_uom.field.choices,
        required=False,
    )
    default_area_uom = serializers.ChoiceField(
        choices=CompanySettings.default_area_uom.field.choices,
        required=False,
    )

    # Fiscal & financial
    fiscal_year_start_month = serializers.IntegerField(
        min_value=1,
        max_value=12,
        required=False,
    )
    fiscal_year_start_day = serializers.IntegerField(
        min_value=1,
        max_value=31,
        required=False,
    )
    fiscal_calendar_type = serializers.ChoiceField(
        choices=CompanySettings.fiscal_calendar_type.field.choices,
        required=False,
    )
    cost_method = serializers.ChoiceField(
        choices=CompanySettings.cost_method.field.choices,
        required=False,
    )
    tax_inclusive_pricing = serializers.BooleanField(required=False)

    # Document defaults
    default_document_language = serializers.CharField(
        max_length=10,
        required=False,
    )
    default_paper_size = serializers.ChoiceField(
        choices=CompanySettings.default_paper_size.field.choices,
        required=False,
    )
    text_direction = serializers.ChoiceField(
        choices=CompanySettings.text_direction.field.choices,
        required=False,
    )

File: frontend/src/app/[locale]/(app)/(org)/settings/_types/company-settings.ts
import { z } from "zod";

// ── Choice value unions ─────────────────────────────────────────────────────

export const DECIMAL_SEPARATORS = ["dot", "comma"] as const;
export const THOUSANDS_SEPARATORS = [
  "comma",
  "dot",
  "space",
  "apostrophe",
  "none",
] as const;
export const DIGIT_GROUPINGS = ["standard", "indian", "none"] as const;
export const CURRENCY_POSITIONS = ["before", "after"] as const;
export const NEGATIVE_FORMATS = [
  "minus_before",
  "minus_after",
  "parentheses",
] as const;
export const ROUNDING_METHODS = [
  "half_up",
  "half_even",
  "ceiling",
  "floor",
] as const;
export const DATE_FORMATS = [
  "YYYY-MM-DD",
  "DD/MM/YYYY",
  "MM/DD/YYYY",
  "DD.MM.YYYY",
  "YYYY/MM/DD",
  "DD-MM-YYYY",
] as const;
export const TIME_FORMATS = ["24h", "12h"] as const;
export const WEEK_START_DAYS = ["monday", "sunday", "saturday"] as const;
export const CALENDAR_SYSTEMS = [
  "gregorian",
  "hijri",
  "thai_buddhist",
  "persian",
  "japanese",
] as const;
export const MEASUREMENT_SYSTEMS = ["metric", "imperial"] as const;
export const WEIGHT_UOMS = ["kg", "g", "mg", "t", "lb", "oz"] as const;
export const LENGTH_UOMS = ["m", "cm", "mm", "ft", "in", "yd"] as const;
export const VOLUME_UOMS = ["L", "mL", "m3", "gal", "fl_oz", "ft3"] as const;
export const TEMPERATURE_UOMS = ["celsius", "fahrenheit"] as const;
export const AREA_UOMS = ["m2", "ft2", "hectare", "acre"] as const;
export const COST_METHODS = ["standard", "average", "fifo"] as const;
export const FISCAL_CALENDAR_TYPES = [
  "standard",
  "445_week",
  "454_week",
  "544_week",
] as const;
export const PAPER_SIZES = ["A4", "letter", "legal"] as const;
export const TEXT_DIRECTIONS = ["ltr", "rtl"] as const;

// ── Zod schema ──────────────────────────────────────────────────────────────

const precision = z.number().int().min(0).max(10);

export const companySettingsSchema = z
  .object({
    // Number formatting
    decimal_separator: z.enum(DECIMAL_SEPARATORS),
    thousands_separator: z.enum(THOUSANDS_SEPARATORS),
    digit_grouping: z.enum(DIGIT_GROUPINGS),

    // Precision
    quantity_precision: precision,
    price_precision: precision,
    currency_precision: precision,
    exchange_rate_precision: precision,
    percentage_precision: precision,
    weight_precision: precision,

    // Currency display
    currency_symbol_position: z.enum(CURRENCY_POSITIONS),
    currency_spacing: z.boolean(),
    negative_format: z.enum(NEGATIVE_FORMATS),

    // Rounding
    rounding_method: z.enum(ROUNDING_METHODS),
    cash_rounding_enabled: z.boolean(),
    cash_rounding_increment: z.string(),

    // Date & time
    date_format: z.enum(DATE_FORMATS),
    time_format: z.enum(TIME_FORMATS),
    week_start_day: z.enum(WEEK_START_DAYS),
    calendar_system: z.enum(CALENDAR_SYSTEMS),

    // Measurement
    measurement_system: z.enum(MEASUREMENT_SYSTEMS),
    default_weight_uom: z.enum(WEIGHT_UOMS),
    default_length_uom: z.enum(LENGTH_UOMS),
    default_volume_uom: z.enum(VOLUME_UOMS),
    default_temperature_uom: z.enum(TEMPERATURE_UOMS),
    default_area_uom: z.enum(AREA_UOMS),

    // Fiscal & financial
    fiscal_year_start_month: z.number().int().min(1).max(12),
    fiscal_year_start_day: z.number().int().min(1).max(31),
    fiscal_calendar_type: z.enum(FISCAL_CALENDAR_TYPES),
    cost_method: z.enum(COST_METHODS),
    tax_inclusive_pricing: z.boolean(),

    // Document defaults
    default_document_language: z.string().min(2).max(10),
    default_paper_size: z.enum(PAPER_SIZES),
    text_direction: z.enum(TEXT_DIRECTIONS),
  })
  .refine(
    (data) => {
      if (
        data.decimal_separator === "dot" &&
        data.thousands_separator === "dot"
      )
        return false;
      if (
        data.decimal_separator === "comma" &&
        data.thousands_separator === "comma"
      )
        return false;
      return true;
    },
    { path: ["thousands_separator"], message: "separator_conflict" },
  )
  .refine(
    (data) => {
      if (data.digit_grouping !== "none" && data.thousands_separator === "none")
        return false;
      return true;
    },
    { path: ["digit_grouping"], message: "digit_grouping_requires_separator" },
  );

// ── Types ───────────────────────────────────────────────────────────────────

export type CompanySettings = z.infer<typeof companySettingsSchema>;

export type CompanySettingsResponse = CompanySettings & {
  created_at: string;
  updated_at: string;
};
